src/Components/bookHub/BookHub.js

```diff
@@ -14,20 +17,77 @@ const base = new Airtable({ apiKey: airtableConfig.apiKey })
 
 function CardsDisplay() {
   const [cards, setCards] = useState([]);
+  const [filteredCards, setFilteredCards] = useState([]);
+  const [rangeInput, setRangeInput] = useState({
+    age: { min: ageRangeMetadata[0], max: ageRangeMetadata[18] },
+    grade: { min: gradeRangeMetadata[0], max: gradeRangeMetadata[12] },
+  });
+  const [multiSelectInput, setMultiSelectInput] = useState({
+    Ethnicity: [],
+    Religion: [],
+    Gender: [],
+    Sexuality: [],
+  });
 
   const getCards = () => {
     base('Book').select({ view: 'Grid view' }).all()
       .then((records) => {
         setCards(records);
+        setFilteredCards(records);
       });
   };
 
-  useEffect(getCards, []);
+  useEffect(() => {
+    const validGradeTags = gradeRangeMetadata.slice(
+      gradeRangeMetadata.indexOf(rangeInput.grade.min),
+      gradeRangeMetadata.indexOf(rangeInput.grade.max) + 1,
+    );
+    const validAgeTags = ageRangeMetadata.slice(
+      ageRangeMetadata.indexOf(rangeInput.age.min),
+      ageRangeMetadata.indexOf(rangeInput.age.max) + 1,
+    );
+    console.log(rangeInput);
+    console.log(multiSelectInput);
+    console.log(cards);
+
+    setFilteredCards(cards.filter(
+      (record) => (record.fields.age_range.some((val) => validAgeTags.indexOf(val) !== -1)
+      && record.fields.grade_range.some((value) => validGradeTags.indexOf(value) !== -1))
+      && (multiSelectInput.Ethnicity.length === 0
```

Comment:
now that you know the correct logic to run for each filter, how can you rewrite this to be slightly more code efficient? If you have time try out the forEach method!